servers/rendering/renderer_rd/shaders/scene_forward_lights_inc.glsl

```diff
@@ -59,200 +59,174 @@ void light_compute(vec3 N, vec3 L, vec3 V, float A, vec3 light_color, bool is_di
 
 	vec2 read_viewport_size = scene_data_block.data.viewport_size;
-
 	vec3 normal = N;
 	vec3 light = L;
 	vec3 view = V;
 
 #CODE : LIGHT
+#else // !LIGHT_CODE_USED
+	float NdotL = min(A + dot(N, L), 1.0);
+	float cNdotV = max(dot(N, V), 1e-4);
 
+#ifdef LIGHT_TRANSMITTANCE_USED
+	{
+#ifdef SSS_MODE_SKIN
+		float scale = 8.25 / transmittance_depth;
+		float d = scale * abs(transmittance_z);
+		float dd = -d * d;
+		vec3 profile = vec3(0.233, 0.455, 0.649) * exp(dd / 0.0064) +
+				vec3(0.1, 0.336, 0.344) * exp(dd / 0.0484) +
+				vec3(0.118, 0.198, 0.0) * exp(dd / 0.187) +
+				vec3(0.113, 0.007, 0.007) * exp(dd / 0.567) +
+				vec3(0.358, 0.004, 0.0) * exp(dd / 1.99) +
+				vec3(0.078, 0.0, 0.0) * exp(dd / 7.41);
+
+		diffuse_light += profile * transmittance_color.a * light_color * clamp(transmittance_boost - NdotL, 0.0, 1.0) * (1.0 / M_PI);
 #else
 
-	float NdotL = min(A + dot(N, L), 1.0);
-	float cNdotL = max(NdotL, 0.0); // clamped NdotL
-	float NdotV = dot(N, V);
-	float cNdotV = max(NdotV, 1e-4);
-
-#if defined(DIFFUSE_BURLEY) || defined(SPECULAR_SCHLICK_GGX) || defined(LIGHT_CLEARCOAT_USED)
-	vec3 H = normalize(V + L);
+		float scale = 8.25 / transmittance_depth;
+		float d = scale * abs(transmittance_z);
+		float dd = -d * d;
+		diffuse_light += exp(dd) * transmittance_color.rgb * transmittance_color.a * light_color * clamp(transmittance_boost - NdotL, 0.0, 1.0) * (1.0 / M_PI);
 #endif
+	}
+#endif //LIGHT_TRANSMITTANCE_USED
 
-#if defined(SPECULAR_SCHLICK_GGX)
-	float cNdotH = clamp(A + dot(N, H), 0.0, 1.0);
+#if defined(LIGHT_RIM_USED)
+	// Epsilon min to prevent pow(0, 0) singularity which results in undefined behavior.
+	float rim_light = pow(max(1e-4, 1.0 - cNdotV), max(0.0, (1.0 - roughness) * 16.0));
+	diffuse_light += rim_light * rim * mix(vec3(1.0), albedo, rim_tint) * light_color;
 #endif
 
+	// We skip checking on attenuation on directional lights to avoid a branch that is not as beneficial for directional lights as the other ones.
+	const float EPSILON = 1e-3f;
+	if (is_directional || attenuation > EPSILON) {
+		float cNdotL = max(NdotL, 0.0);
+#if defined(DIFFUSE_BURLEY) || defined(SPECULAR_SCHLICK_GGX) || defined(LIGHT_CLEARCOAT_USED)
+		vec3 H = normalize(V + L);
+#endif
 #if defined(DIFFUSE_BURLEY) || defined(SPECULAR_SCHLICK_GGX) || defined(LIGHT_CLEARCOAT_USED)
-	float cLdotH = clamp(A + dot(L, H), 0.0, 1.0);
+		float cLdotH = clamp(A + dot(L, H), 0.0, 1.0);
 #endif
+#if defined(LIGHT_CLEARCOAT_USED)
+		// Clearcoat ignores normal_map, use vertex normal instead
+		float ccNdotL = max(min(A + dot(vertex_normal, L), 1.0), 0.0);
+		float ccNdotH = clamp(A + dot(vertex_normal, H), 0.0, 1.0);
+		float ccNdotV = max(dot(vertex_normal, V), 1e-4);
+
+#if !defined(SPECULAR_SCHLICK_GGX)
+		float cLdotH5 = SchlickFresnel(cLdotH);
+#endif
+		float Dr = D_GGX(ccNdotH, mix(0.001, 0.1, clearcoat_roughness));
+		float Gr = 0.25 / (cLdotH * cLdotH);
+		float Fr = mix(.04, 1.0, cLdotH5);
```

Comment:
Pretty sure this is the source of the errors I'm getting on a fresh compile:
```
  servers\rendering\renderer_rd\shader_rd.cpp:336 - Error compiling Fragment shader, variant #16 ().
  servers\rendering\renderer_rd\shader_rd.cpp:337 - Failed parse:
  WARNING: 0:60: '' : all default precisions are highp; use precision statements to quiet warning, e.g.:
           "precision mediump int; precision highp float;" 
  ERROR: 0:1086: 'cLdotH5' : undeclared identifier 
  ERROR: 0:1086: '' : missing #endif 
  ERROR: 0:1086: '' : compilation terminated 
  ERROR: 3 compilation errors.  No code generated.
```

`cLdotH5` is only defined when `!defined(SPECULAR_SCHLICK_GGX)` is `true`. Should this section be moved into the same conditional block, or should `cLdotH5` be defined outside of it?

Reply:
We should remove the `!defined(SPECULAR_SCHLICK_GGX)` check in the clearcoat section. and then in the specular GGX section add a `!defined(LIGHT_CLEARCOAT_USED)` and only define `cLdotH5` when clearcoat hasn't already defined it

Reply:
Fixed in https://github.com/godotengine/godot/pull/100081